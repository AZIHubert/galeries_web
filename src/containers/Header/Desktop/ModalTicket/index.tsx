import { useFormik } from 'formik';
import * as React from 'react';
import {
  useDispatch,
  useSelector,
} from 'react-redux';

import Field from '#components/Field';
import Button from '#components/Button';
import Modal from '#components/Modal';
import RequiredField from '#components/RequiredField';
import Text from '#components/Text';

import { ticketSchema } from '#helpers/schemas';

import {
  fetchSendTicket,
  resetSendTicket,
  setSendTicket,
} from '#store/actions';
import {
  loadingSelector,
  sendTicketErrorSelector,
  sendTicketStatusSelector,
} from '#store/selectors';

const initialValues: form.SendTicketI = {
  body: '',
  header: '',
};

interface ModalTicketI {
  handleClose: () => void;
}

const ModalTicket = ({
  handleClose,
}: ModalTicketI) => {
  const dispatch = useDispatch();
  const formik = useFormik({
    initialValues,
    onSubmit: (value) => {
      if (!loading) {
        dispatch(fetchSendTicket(value));
      }
    },
    validateOnBlur: true,
    validateOnChange: false,
    validationSchema: ticketSchema,
  });
  const loading = useSelector(loadingSelector);
  const sendTicketError = useSelector(sendTicketErrorSelector);
  const sendTicketStatus = useSelector(sendTicketStatusSelector);

  React.useEffect(() => () => {
    dispatch(resetSendTicket());
  }, []);
  React.useEffect(() => {
    if (sendTicketStatus === 'success') {
      handleClose();
    }
  }, [sendTicketStatus]);

  return (
    <Modal.Container>
      <Text
        color='primary'
        styles={{
          fontSize: 1.2,
          marginBottom: 20,
          textAlign: 'center',
        }}
        stylesMobile={{
          fontSize: 1.4,
        }}
        stylesLaptopL={{
          fontSize: 1.5,
        }}
      >
        Create galerie
      </Text>
      <form
        data-testid='form'
        onSubmit={formik.handleSubmit}
      >
        <Field
          disabled={loading}
          error={
            formik.errors.header || sendTicketError.header
          }
          fieldTestId='header'
          id='header'
          label='title'
          onBlur={formik.handleBlur}
          onChange={(e) => {
            formik.handleChange(e);
            if (sendTicketError.header) {
              dispatch(setSendTicket({
                errors: {
                  ...sendTicketError,
                  header: '',
                },
              }));
            }
          }}
          required
          styles={{
            marginBottom: 6,
          }}
          stylesLaptopL={{
            marginBottom: 10,
          }}
          touched={formik.touched.header}
          value={formik.values.header}
        />
        <Field
          disabled={loading}
          error={
            formik.errors.body || sendTicketError.body
          }
          fieldTestId='body'
          id='body'
          multiline
          label='body'
          onBlur={formik.handleBlur}
          onChange={(e) => {
            formik.handleChange(e);
            if (sendTicketError.body) {
              dispatch(setSendTicket({
                errors: {
                  ...sendTicketError,
                  body: '',
                },
              }));
            }
          }}
          required
          styles={{
            marginBottom: 12,
          }}
          stylesLaptopL={{
            marginBottom: 15,
          }}
          touched={formik.touched.body}
          value={formik.values.body}
        />
        <RequiredField />
        <Button.Gradiant
          disabled={loading}
          styles={{
            marginBottom: 15,
            marginTop: 15,
          }}
          stylesLaptopL={{
            marginBottom: 22,
            marginTop: 22,
          }}
          type='submit'
          title='send'
        />
      </form>
    </Modal.Container>
  );
};

export default ModalTicket;
