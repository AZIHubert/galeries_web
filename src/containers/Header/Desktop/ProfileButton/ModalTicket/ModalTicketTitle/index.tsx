import * as React from 'react';

import Text from '#components/Text';

import {
  Container,
} from './styles';

const ModalTicketTitle = () => (
  <Container>
    <Text
      color='primary'
      styles={{
        fontSize: 1.2,
        textAlign: 'center',
      }}
      stylesMobile={{
        fontSize: 1.4,
      }}
      stylesLaptopL={{
        fontSize: 1.5,
      }}
    >
      Send a ticket
    </Text>
  </Container>
);

export default ModalTicketTitle;
