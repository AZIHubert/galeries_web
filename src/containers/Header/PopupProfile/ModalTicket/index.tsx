import * as React from 'react';
import { useFormik } from 'formik';

import { ticketSchema } from '#helpers/schemas';

interface ModalTicketI {
  loading: boolean;
  setLoading: React.Dispatch<React.SetStateAction<boolean>>;
}

const initialValues = {
  header: '',
  body: '',
};

const ModalTicket = ({ loading, setLoading }: ModalTicketI) => {
  const formik = useFormik({
    initialValues,
    onSubmit: () => {
      if (!loading) { setLoading(true); }
    },
    validateOnChange: false,
    validateOnBlur: true,
    validationSchema: ticketSchema,
  });
  return (
    <div>
      <form onSubmit={formik.handleSubmit}>
        <label
          htmlFor="header"
        >
          title *
        </label>
        <input
          data-testid='headerField'
          disabled={loading}
          id='header'
          name='header'
          onBlur={formik.handleBlur}
          onChange={formik.handleChange}
          type='text'
          value={formik.values.header}
        />
        {formik.errors.header
        && formik.touched.header
        && (
          <div
            data-testid='headerError'
          >
            {formik.errors.header}
          </div>
        )}
        <label
          htmlFor="body"
        >
          body *
        </label>
        <input
          data-testid='bodyField'
          disabled={loading}
          id='body'
          name='body'
          onBlur={formik.handleBlur}
          onChange={formik.handleChange}
          type='text'
          value={formik.values.body}
        />
        {formik.errors.body && formik.touched.body && (
          <div
            data-testid='bodyError'
          >
            {formik.errors.body}
          </div>
        )}
        <div>
          * Required field
        </div>
        <button
          disabled={loading}
          data-testid='submitButton'
          type='submit'
        >
          {loading ? 'loading' : 'send'}
        </button>
      </form>
    </div>
  );
};

export default ModalTicket;
