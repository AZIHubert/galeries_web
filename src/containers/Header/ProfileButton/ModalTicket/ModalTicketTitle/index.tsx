import * as React from 'react';
import styled from 'styled-components';

import Text from '#components/Text';

const Container = styled.div`
  margin-bottom: 20px;
`;

const ModalTicketTitle = () => (
  <Container>
    <Text
      textAlign='center'
      fontSize={1.4}
    >
      Send a ticket
    </Text>
  </Container>
);

export default ModalTicketTitle;
