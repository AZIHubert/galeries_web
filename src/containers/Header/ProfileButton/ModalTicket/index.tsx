import { useFormik } from 'formik';
import * as React from 'react';
import {
  useSelector,
  useDispatch,
} from 'react-redux';

import Field from '#components/Field';
import GradientButton from '#components/GradientButton';
import ModalContainer from '#components/ModalContainer';
import RequiredField from '#components/RequiredField';

import { ticketSchema } from '#helpers/schemas';

import {
  fetchSendTicket,
  resetSendTicket,
  setSendTicket,
} from '#store/actions';
import {
  loadingSelector,
  sendTicketErrorSelector,
  sendTicketStatusSelector,
} from '#store/selectors';

const initialValues: form.SendTicketI = {
  body: '',
  header: '',
};

interface ModalTicketI {
  handleClose: () => void;
}

const ModalTicket = ({
  handleClose,
}: ModalTicketI) => {
  const dispatch = useDispatch();
  const formik = useFormik({
    initialValues,
    onSubmit: (value) => {
      if (!loading) {
        resetForm();
        dispatch(fetchSendTicket(value));
      }
    },
    validateOnBlur: true,
    validateOnChange: false,
    validationSchema: ticketSchema,
  });
  const loading = useSelector(loadingSelector);
  const sendTicketError = useSelector(sendTicketErrorSelector);
  const sendTicketStatus = useSelector(sendTicketStatusSelector);

  React.useEffect(() => () => resetForm(), []);
  React.useEffect(() => {
    if (sendTicketStatus === 'success') {
      handleClose();
    }
  }, [sendTicketStatus]);

  const resetForm = () => {
    dispatch(resetSendTicket());
  };

  return (
    <ModalContainer
      title='Send a ticket'
      titleMarginTop={20}
      titleTextAlign='center'
    >
      <form
        data-testid='form'
        onSubmit={formik.handleSubmit}
      >
        <Field
          disabled={loading}
          fieldTestId='header'
          error={
            formik.errors.header || sendTicketError.header
          }
          id='header'
          marginBottom={6}
          marginBottomL={10}
          label='title'
          onBlur={formik.handleBlur}
          onChange={(e) => {
            formik.handleChange(e);
            if (sendTicketError.header) {
              dispatch(setSendTicket({
                errors: {
                  ...sendTicketError,
                  header: '',
                },
              }));
            }
          }}
          required
          touched={formik.touched.header}
          value={formik.values.header}
        />
        <Field
          disabled={loading}
          error={
            formik.errors.body || sendTicketError.body
          }
          fieldTestId='body'
          id='body'
          marginBottom={12}
          marginBottomL={15}
          multiline
          label='body'
          onBlur={formik.handleBlur}
          onChange={(e) => {
            formik.handleChange(e);
            if (sendTicketError.body) {
              dispatch(setSendTicket({
                errors: {
                  ...sendTicketError,
                  body: '',
                },
              }));
            }
          }}
          required
          touched={formik.touched.body}
          value={formik.values.body}
        />
        <RequiredField />
        <GradientButton
          disabled={loading}
          marginBottom={15}
          marginBottomL={22}
          marginTop={15}
          marginTopL={22}
          type='submit'
          title='send'
        />
      </form>
    </ModalContainer>
  );
};

export default ModalTicket;
