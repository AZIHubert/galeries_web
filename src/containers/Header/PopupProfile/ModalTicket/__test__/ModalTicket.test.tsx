import * as React from 'react';
import renderer from 'react-test-renderer';
import {
  cleanup,
  fireEvent,
  render,
  screen,
} from '@testing-library/react';

import {
  MAX_LENGTH_THRITY,
  MAX_LENGTH_TWO_HUNDRER,
  MIN_LENGTH_OF_FIVE,
  MIN_LENGTH_OF_TEN,
  REQUIRED,
} from '#helpers/formErrors';

import ModalTicket from '../index';

const Container = () => {
  const [loading, setLoading] = React.useState(false);
  return (
    <ModalTicket
      loading={loading}
      setLoading={setLoading}
    />
  );
};

describe('ModalSignin', () => {
  const setLoading = jest.fn;
  let bodyField: HTMLElement;
  let headerField: HTMLElement;
  let submitButton: HTMLElement;
  beforeEach(() => {
    const { getByTestId } = render(<Container />);
    bodyField = getByTestId('bodyField');
    headerField = getByTestId('headerField');
    submitButton = getByTestId('submitButton');
  });
  afterEach(cleanup);
  it('renders without crashing', () => {
    const tree = renderer.create(<ModalTicket loading={false} setLoading={setLoading} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('should not loading if errors', async () => {
    fireEvent.click(submitButton);
    const headerError = await screen.findByTestId('headerError');
    const bodyError = await screen.findByTestId('bodyError');
    expect(headerField).not.toBeDisabled();
    expect(headerError).toHaveTextContent(REQUIRED);
    expect(submitButton).toHaveTextContent('send');
    expect(submitButton).not.toBeDisabled();
    expect(headerField).not.toBeDisabled();
    expect(bodyError).toHaveTextContent(REQUIRED);
  });
  it('should loading if no error', async () => {
    fireEvent.change(headerField, { target: { value: 'a'.repeat(5) } });
    fireEvent.change(bodyField, { target: { value: 'a'.repeat(10) } });
    fireEvent.click(submitButton);
    await screen.findByText('loading');
    expect(bodyField).toBeDisabled();
    expect(headerField).toBeDisabled();
    expect(submitButton).toBeDisabled();
  });
  describe('should return an error if', () => {
    describe('header', () => {
      it('is empty', async () => {
        fireEvent.blur(headerField);
        const headerError = await screen.findByTestId('headerError');
        expect(headerError).not.toBeNull();
        expect(headerError).toHaveTextContent(REQUIRED);
      });
      it('is less than 5 characters', async () => {
        fireEvent.change(headerField, { target: { value: 'a'.repeat(4) } });
        fireEvent.blur(headerField);
        const headerError = await screen.findByTestId('headerError');
        expect(headerError).not.toBeNull();
        expect(headerError).toHaveTextContent(MIN_LENGTH_OF_FIVE);
      });
      it('is more than 50 characters', async () => {
        fireEvent.change(headerField, { target: { value: 'a'.repeat(31) } });
        fireEvent.blur(headerField);
        const headerError = await screen.findByTestId('headerError');
        expect(headerError).not.toBeNull();
        expect(headerError).toHaveTextContent(MAX_LENGTH_THRITY);
      });
    });
    describe('body', () => {
      it('is empty', async () => {
        fireEvent.blur(bodyField);
        const bodyError = await screen.findByTestId('bodyError');
        expect(bodyError).not.toBeNull();
        expect(bodyError).toHaveTextContent(REQUIRED);
      });
      it('is less than 10 characters', async () => {
        fireEvent.change(bodyField, { target: { value: 'a'.repeat(9) } });
        fireEvent.blur(bodyField);
        const bodyError = await screen.findByTestId('bodyError');
        expect(bodyError).not.toBeNull();
        expect(bodyError).toHaveTextContent(MIN_LENGTH_OF_TEN);
      });
      it('is more than 200 characters', async () => {
        fireEvent.change(bodyField, { target: { value: 'a'.repeat(201) } });
        fireEvent.blur(bodyField);
        const bodyError = await screen.findByTestId('bodyError');
        expect(bodyError).not.toBeNull();
        expect(bodyError).toHaveTextContent(MAX_LENGTH_TWO_HUNDRER);
      });
    });
  });
});
