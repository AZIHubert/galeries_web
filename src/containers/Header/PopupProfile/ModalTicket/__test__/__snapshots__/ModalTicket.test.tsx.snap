// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalSignin renders without crashing 1`] = `
<div
  data-testid="ticketModal"
>
  <form
    onSubmit={[Function]}
  >
    <label
      htmlFor="header"
    >
      title *
    </label>
    <input
      data-testid="headerField"
      disabled={false}
      id="header"
      name="header"
      onBlur={[Function]}
      onChange={[Function]}
      type="text"
      value=""
    />
    <label
      htmlFor="body"
    >
      body *
    </label>
    <input
      data-testid="bodyField"
      disabled={false}
      id="body"
      name="body"
      onBlur={[Function]}
      onChange={[Function]}
      type="text"
      value=""
    />
    <div>
      * Required field
    </div>
    <button
      data-testid="submitButton"
      disabled={false}
      type="submit"
    >
      send
    </button>
  </form>
</div>
`;
